### PR TITLE
services/horizon/internal/actions: Add latest ingested ledger in experimental ingestion endpoints

### PR DESCRIPTION
**Pull request by tamirms**
Include a Latest-Ledger header with the sequence number of the last processed ledger by the
experimental ingestion system. Also, ensure the endpoint responses contains only data consistent
with the sequence number included in the header

<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [x] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### Summary

Add `Latest-Ledger` header with the sequence number of the last processed ledger by the experimental ingestion system. The endpoints built using the experimental ingestion system will always respond with data which is consistent with the ledger in `Latest-Ledger`.

Fixes https://github.com/stellar/go/issues/1791

### Goal and scope

See https://github.com/stellar/go/issues/1791 for motivation

### CODE REVIEW

File: services/horizon/internal/actions/offer.go [hunk 1]
```diff
@@ -58,7 +62,6 @@ func (handler GetOffersHandler) GetResourcePage(r *http.Request) ([]hal.Pageable
 // GetAccountOffersHandler is the action handler for the
 // `/accounts/{account_id}/offers` endpoint when using experimental ingestion.
 type GetAccountOffersHandler struct {
-	HistoryQ *history.Q
```
Comment (abuiles) [inline services/horizon/internal/actions/offer.go]: what's the rationale for this change? that we can access Q from context?

Comment (abuiles) [reply services/horizon/internal/actions/offer.go]: what was the reason for not doing it this way initially?

Comment (tamirms) [reply services/horizon/internal/actions/offer.go]: we need to create a repeatable read transaction to ensure that all the data in the handler's response is consistent with the sequence number we will set in the response header.

there are 2 ways we can do this. we can create a helper function to create the repeatable read transaction. then we would call the helper function in each of the handlers. we'd also need to make sure the transaction is rolled back once the handler finishes executing.

alternatively, we could create this transaction and handle the rollback in a middleware. then we can inject the transaction in the request's context.

the reason we chose not to inject  `history.Q` into the request's context is because we wanted to avoid relying upon the app context since that was one of the stated goals of the actions refactoring.

In this PR, I have decided to inject the transaction session to make sure the repeatable read transaction is constructed in one place. However, I would not be opposed to walking back that decision and taking the alternative route

**Comment by tamirms** [comment]
@bartekn @abuiles this is ready for review now

File: exp/orderbook/batch.go [hunk 2]
```diff
@@ -60,6 +60,10 @@ func (tx *orderBookBatchedUpdates) apply() error {
 	}
 	tx.committed = true
 
+	if ledger < tx.orderbook.lastLedger {
+		return errOutdatedLedger
+	}
```
Comment (bartekn) [inline exp/orderbook/batch.go]: We should probably check: `tx.orderbook.lastLedger != 0 && ledger != tx.orderbook.lastLedger + 1` to make sure no ledger is skipped.

File: services/horizon/internal/handler.go [hunk 3]
```diff
@@ -342,18 +342,24 @@ func validateCursorWithinHistory(pq db2.PageQuery) error {
 const singleObjectStreamLimit = 10
 
 type streamableObjectAction interface {
-	GetResource(r *http.Request) (actions.StreamableObjectResponse, error)
+	GetResource(
+		r *http.Request,
+		w actions.HeaderWriter,
```
Comment (bartekn) [inline services/horizon/internal/handler.go]: What do you think about flipping the order so it matches `ServeHTTP`?

File: services/horizon/internal/paths/main.go [hunk 4]
```diff
@@ -28,7 +28,7 @@ type Path struct {
 // Finder finds paths.
 type Finder interface {
 	// Returns path for a Query of a maximum length `maxLength`
-	Find(q Query, maxLength uint) ([]Path, error)
+	Find(q Query, maxLength uint) ([]Path, uint32, error)
```
Comment (bartekn) [inline services/horizon/internal/paths/main.go]: Please update the comment to explain what returned `uint32` is.
